Feature/auth

vedantroy opened the pull request:
This PR adds basic username + password authentication. It includes
- Routes to login and logout
- A route `/auth/test` that is an example "protected" route (cannot be accessed unless the user is logged in).

It does not have a route to add new accounts to the database because I figured there will probably only be 1 or 2 admin users, so there's no need to have a API route for adding new accounts.

Notes:
- Almost all the authentication is handled using third party libraries, such as `passport` and `bcrypt`. This reduces our room to make mistakes.
    - `bcrypt` handles the hashing + salting of passwords. It's a well-known, trustworthy library.
    - `passport` handles user sessions. Also well-known/widely used.
- I did not add the changes to the dev server file to the production server file because there would be a lot of code duplication. Not sure what to do about that. Maybe I can extract the backend server code into a separate file because only the front-end server code changes between dev and prod (due to the use of webpack middleware in development).

vedantroy committed:
Ignore sessions folder created by session-file-store

vedantroy committed:
Install dependencies for auth

vedantroy committed:
Refactor manage postgres script and add support for users table

vedantroy committed:
Add auth routes

vedantroy committed:
Configure passport.js, express-session, and auth route in dev server

stevenlaitran commented on gifted_build/src/server/dev-server.js:
```diff
@@ -25,6 +29,23 @@ app.use('*', (req, res, next) => {
+server.use(bodyParser.json())
 server.use(require('./routes/products'));
 
+server.use(session({
```
We should set up a "common_configs" folder that both the dev-server and prod-server can reference.

stevenlaitran reviewed (changes requested):
auth looks good! creating a config routes file is the only thing I would change

vedantroy committed:
Extract common initialization into separate file

vedantroy committed:
Add fake production DB config

vedantroy committed:
Rename server files to dev and prod

vedantroy commented:
@stevenlaitran Sounds good. I extracted the common configuration into a separate file.

stevenlaitran commented on gifted_build/src/server/start_servers.js:
```diff
@@ -0,0 +1,60 @@
+const path = require('path')
+
+module.exports.startServers = mode => {
+    if (mode === "DEVELOPMENT") {
```
This should not be placed in a common express file. 

Development and production builds require different dependencies and packages, and the import commands in a common file will require npm to install all in order to run them. We do not want webpack-dev-middleware/webpack-hot-middleware to be required when in production mode.

I highly recommend elevating these commands back into separate dev and prod files respectively.

stevenlaitran commented on gifted_build/package.json:
```diff
@@ -7,7 +7,7 @@
   "scripts": {
     "test": "webpack -v",
     "build": "webpack --config webpack-prod.config.js && webpack --config webpack-server.config.js",
-    "start": "node src/server/dev-server.js",
```
Is there a reason you renamed these files? This might negate other work done on branches, but since not many others are working in the backend, this shouldn't affect too much, although naming references need to change.

vedantroy replied:
I had no real reason to rename the files besides making it a bit clearer to myself that `server.js` was the production server. I can undo the renaming, if it makes things easier/clearer.

stevenlaitran commented on gifted_build/package.json:
```diff
@@ -7,7 +7,7 @@
     "build": "webpack --config webpack-prod.config.js && webpack --config webpack-server.config.js",
-    "start": "node src/server/dev-server.js",
+    "start": "node src/server/dev.js",
     "start:prod": "node dist/server.js"
```
This should be now 
"start:prod": "node dist/prod.js"

with the new naming scheme

stevenlaitran commented on gifted_build/src/server/start_servers.js:
```diff
@@ -0,0 +1,60 @@
+    server.use(bodyParser.json())
+    server.use(require('./routes/products'));
+
+    server.use(session({
```
I assume this should be different between dev and prod environments as the secret key should be changed regularly in prod as opposed to the dev environment.

stevenlaitran reviewed (changes requested):
A few changes regarding the renaming of the express files and changes in the environments

vedantroy committed:
Refactor dev and prod server configuration

vedantroy committed:
Finish updating names of dev and production server files

vedantroy commented:
@stevenlaitran 

I refactored the server configuration/startup files.

I also finished renaming the files to `dev.js` and `prod.js` and updated the `package.json`. I don't think the other backend developer has made a branch that would conflict with this renaming, so the renaming shouldn't cause a hassle, but I'm also fine with changing back the names to their original of `dev-server` and `server`.

I tested `npm run start` and `npm run start:prod`, and both seem to be working correctly with the new naming scheme.

stevenlaitran committed:
fixed listen on port 3000 and added environment variable for database configs

stevenlaitran commented:
@vedantroy I noticed you deleted the listen on port 3000 for both dev and prod server files and I just decided to go ahead and fix it. In addition, I changed the hierarchy for database commands and added to the pgnote file, with support for configs to be stored in a separate file that can be specified when either in a dev/prod environment. Dev will always refer to the database.ini file that is in the database directory. 

Once you look it over, I'll merge this. Thanks for the hard work!

vedantroy committed:
Update database setup instructions.

The database "api" needs to be created under the user "build". This
requires logging into postgres as the "build" user.

vedantroy commented:
Looks good! I just updated the database setup instructions a bit. @stevenlaitran
